### PR TITLE
gh-120754: Ensure _stat_atopen is cleared on fd change

### PR DESCRIPTION
**Pull request by cmaloney**
While adding _stat_atopen I accidentally created a leak in the C implementation because the `_stat_atopen` member was set to NULL in the C implementation, but the memory wasn't freed. That was resolved, but it showed there were a number of potential cases where `_stat_atopen` _should_ be cleared (the information is out of date because the fd changed) that it was not. In this PR I audited by reading through FileIO (in _io and _pyio), and in all cases where the fd is changed/closed made sure to clear/reset _stat_atopen as well.

Followup from: https://github.com/python/cpython/pull/124225#issuecomment-2359574937, cc: @vstinner 

* Issue: gh-120754

### CODE REVIEW

File: Modules/_io/fileio.c [hunk 1]
```diff
@@ -267,9 +272,12 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
 #endif
     if (self->fd >= 0) {
         if (self->closefd) {
-            /* Have to close the existing file first. */
-            if (internal_close(self) < 0)
+            /* Have to close the existing file first.
+
+               This leaves the stat so we can reuse the memory. */
```
Comment (vstinner) [inline Modules/_io/fileio.c]: I don't think that this optimization is worth it, I would prefer to always clear the cache in close().

File: Modules/_io/fileio.c [hunk 2]
```diff
@@ -455,11 +458,17 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
 #endif
     }
 
-    PyMem_Free(self->stat_atopen);
```
Comment (cmaloney) [inline Modules/_io/fileio.c]: Not sure which is preferred here. The Free + New back to back feels weird, but version with the conditional feels like "extra work/code that _should_ never be executed" (but that the compiler / tools that run currently won't validate...).

**Review by vstinner — COMMENTED** [review]
There are compiler warnings that you can see in the review tab.

**Comment by cmaloney** [comment]
I think the latest commit fixes the compiler warnings (was a testing/debugging assert I accidentally committed with a missing `;') latest changes remove that assert). Not able to find them in the latest changes.

File: Modules/_io/fileio.c [hunk 3]
```diff
@@ -523,10 +531,8 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
         internal_close(self);
         _PyErr_ChainExceptions1(exc);
     }
-    if (self->stat_atopen != NULL) {
-        PyMem_Free(self->stat_atopen);
-        self->stat_atopen = NULL;
-    }
+    PyMem_Free(self->stat_atopen);
+    self->stat_atopen = NULL;
```
Comment (vstinner) [inline Modules/_io/fileio.c:530]: You should clear stat_atopen before calling internal_close() to be consistent with the Python implementation.

File: Modules/_io/fileio.c (hunk 2 above)
Comment (vstinner) [reply Modules/_io/fileio.c]: Maybe `if (ptr != NULL) free(ptr)` would be more regular and avoid the 4 lines long comment.

Comment (cmaloney) [reply Modules/_io/fileio.c]: Moved to that, dropped comment and simplified / back to always calling PyMem_New

**Comment by cmaloney** [comment]
CI "Tests / Ubuntu SSL tests with OpenSSL" failed because of what looks like a network hang / not these changes: `Received 164773440 of 168967744 (97.5%), 0.0 MBs/sec` while doing "Configure ccache action"

File: Modules/_io/fileio.c [hunk 4]
```diff
@@ -178,6 +180,8 @@ _io_FileIO_close_impl(fileio *self, PyTypeObject *cls)
             PyErr_Clear();
         }
     }
+    PyMem_Free(self->stat_atopen);
+    self->stat_atopen = NULL;
```
Comment (vstinner) [inline Modules/_io/fileio.c]: internal_close() already does the same, so it's redundant, no?

File: Modules/_io/fileio.c [hunk 5]
```diff
@@ -266,10 +270,13 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
     assert(PyFileIO_Check(state, self));
 #endif
     if (self->fd >= 0) {
+        PyMem_Free(self->stat_atopen);
+        self->stat_atopen = NULL;
```
Comment (vstinner) [inline Modules/_io/fileio.c]: internal_close() already does the same, so it's redundant, no?

File: Modules/_io/fileio.c [hunk 6]
```diff
@@ -455,7 +462,9 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
 #endif
     }
 
-    PyMem_Free(self->stat_atopen);
+    if (self->stat_atopen != NULL) {
```
Comment (vstinner) [inline Modules/_io/fileio.c]: PyMem_Free(NULL) is well defined: it does nothing, so you might omit the `if (stat_atopen != NULL)` test.

File: Modules/_io/fileio.c (hunk 4 above)
Comment (cmaloney) [reply Modules/_io/fileio.c]: Tradeoff for matching what `_pyio` does which does `os.close()` in these cases vs an `internal_close` that does the OS close + other cleanup. Happy to drop though / definitely feels redundant to me in these two cases.

File: Modules/_io/fileio.c (hunk 5 above)
Comment (cmaloney) [reply Modules/_io/fileio.c]: Moderately, in the `self->closefd` is false case behavior would differ from _pyio, but that will be caught by the PyMem_Free + PyMem_New below, so think this is always redundant / not needed

**Comment by vstinner** [comment]
Merged, thank you.

**Pull request merged**